Dashboard: i18n, refs #8161 (1/2)

sevein opened the pull request:
**i18n**-*ize* all the things,

See also:
- Redmine: https://projects.artefactual.com/issues/8161
- Notes: https://gist.github.com/sevein/75732d85e129348dc32e6c4b15982bf8

Tasks:
- [x] Dashboard: templates, views, etc...
- [x] Dashboard: JavaScript code
- [x] Dashboard: appraisal tab (https://github.com/artefactual-labs/appraisal-tab/pull/151)
- [x] Dashboard: transfer browser (https://github.com/artefactual-labs/transfer_browser/pull/12)
- [x] Archivematica FPR (https://github.com/artefactual/archivematica-fpr-admin/pull/32)
- [x] Field `Job.currentstep`
- [x] Dashboard: use new client, DashboardSetting dicts, etc...

sevein commented:
Actually, not sure if this is meant to be done for 1.6 o 1.7... feel free to update!

sevein commented:
Updated models.

Hwesta commented on src/dashboard/src/requirements/base.txt:
```diff
@@ -9,6 +9,7 @@ django-extensions==1.1.1
 django-autoslug==1.7.1
 django-annoying==0.7.7
 elasticsearch>=1.0.0,<2.0.0
+git+https://github.com/artefactual/archivematica-fpr-admin.git@dev/issue-8161-dashboard-i18n#egg=fpr
```
This seems to generate a warning (error?) `Running setup.py (path:/tmp/pip-build-iBYv5A/fpr/setup.py) egg_info for package fpr produced metadata for project name archivematica-fpr-admin. Fix your #egg=fpr fragments`

sevein replied:
I've fixed this, also now pointing to qa/1.x.

Hwesta commented on src/dashboard/src/components/administration/urls.py:
```diff
@@ -29,6 +30,8 @@
     url(r'dips/atk/$', views.administration_atk_dips),
     url(r'dips/atom/$', views.atom_dips),
     url(r'dips/atom/edit_levels/$', views.atom_levels_of_description),
+    url(r'^i18n/', include('django.conf.urls.i18n')),
```
What's the `i18n` url doing? I can't get anything at `administration/i18n/` and `administration/i18n/setlang/` just redirects back to transfers.

sevein replied:
It exposes the named urlpattern `set_language`. I'm POST'ing to it from `language.html`, see the form tag:

``` html
<form action="{% url 'set_language' %}" method="post">
   <!-- ... -->
</form>
```

Hwesta replied:
Ah, okay.

Hwesta commented on src/dashboard/src/components/administration/forms.py:
```diff
@@ -149,30 +156,14 @@ class Meta:
-        help_text="Used in metadata-only DIP upload.")
+    dip_upload_atom_url = forms.CharField(required=True, label=_("Upload URL"), help_text=_("URL where the Qubit index.php frontend lives, SWORD services path will be appended."))
+    dip_upload_atom_email = forms.CharField(required=True, label=_("Login email"), help_text=_("E-mail account used to log into Qubit."))
+    dip_upload_atom_password = forms.CharField(required=True, label=_("Login password"), help_text=_("E-mail account used to log into Qubit."))
```
This is the original text, but 'E-mail account used to log into Qubit.' doesn't make sense as help text on the password.

Hwesta commented on src/dashboard/src/components/administration/forms.py:
```diff
@@ -210,114 +201,114 @@ class ProcessingConfigurationForm(forms.Form):
         'name': 'quarantine_expiry_days',
-        'label': 'Remove from quarantine after (days)',
+        'label': _('Remove from quarantine after (days)'),
         'placeholder': 'days',
```
Should placeholder be translated too?

Hwesta commented on src/dashboard/src/components/administration/models.py:
```diff
@@ -18,37 +21,38 @@ class Meta:
 EAD_ACTUATE_CHOICES = [('none', 'None'), ('onLoad', 'onLoad'), ('other', 'other'), ('onRequest', 'onRequest')]
 EAD_SHOW_CHOICES = [('embed', 'Embed'), ('new', 'New'), ('none', 'None'), ('other', 'Other'), ('replace', 'Replace')]
```
Should these be translated too?

Hwesta commented on src/dashboard/src/main/views.py:
```diff
@@ -22,6 +22,12 @@
+from django.utils.translation import ugettext as _
+from django.views.decorators.cache import cache_page
+from django.views.decorators.http import last_modified
+from django.views.i18n import javascript_catalog
```
Are all these imports used? I only see `_`

sevein replied:
Right, that's going to be later used by the i18njs catalogue view. I'll fix it.

Hwesta commented on src/dashboard/src/templates/administration/term_detail.html:
```diff
@@ -35,7 +31,8 @@
       <div style='float: right'>
-        <input type='submit' value='Save' /> <a href='/administration/taxonomy/term/{{ term.pk }}/delete/' class='btn btn-danger'>Delete</a>
+        <input type='submit' value='{% trans "Save" %}' />
+        <a href='/administration/taxonomy/term/{{ term.pk }}/delete/' class='btn btn-danger'>{% trans "Delete" %}</a>
```
Out of scope for this, but a PR updating all URLs to use `{% url ... %}` should happen at some point.

sevein replied:
Right! I didn't even notice!

Hwesta commented on src/dashboard/src/templates/ingest/backlog/_search_form.html:
```diff
@@ -1,10 +1,12 @@
-      <input type='button' class='form-control btn-primary' id='search_submit' value='Search transfer backlog'>
       {% if show_files %}
-      <span style='margin-left: 10px'>Show files?</span>
+      <input type='button' class='form-control btn-primary' id='search_submit' value='{% trans "Search transfer backlog" %}'>
```
The search transfer backlog button should not be inside `{% if show files %}`

Hwesta commented on src/dashboard/src/templates/ingest/grid.html:
```diff
@@ -180,32 +181,33 @@
     <div class="job-detail-microservice">
-      <span class="job-type-label">Job:&nbsp;</span><span title="<%= uuid %>"><%= type %> <% if(obj.filename) { %>(file: <i><%= filename %></i>)<% } %></span>
+      <span class="job-type-label">{% trans "Job" %}</span>
+      <span title="<%= uuid %>"><%= type %> </span>
```
Is `<% if(obj.filename) { %>(file: <i><%= filename %></i>)<% } %>` unnecessary now?

sevein replied:
That was a mistake!

Hwesta commented on src/dashboard/src/templates/main/tasks.html:
```diff
@@ -11,21 +12,55 @@
-            <span>Created time:</span> <span class='datetime'>{{ item.createdtime.isoformat }}</span><br />
-            <span>Duration:</span> {{ item.duration }} second(s)
+            {% blocktrans with datetime=item.starttime.isoformat %}
+              <span>Start time:</span> <span class="datetime">{{ datetime }}</span>
```
Does the HTML & value need to be included in the translation or just the label?  I suppose it might be necessary for RTL reordering.  Somewhat concerned about putting styling in translations.

sevein replied:
You're right. I'm fixing it. I'm making some rearrangements so the different properties are described using `<dl></dl>`.

Hwesta commented on src/dashboard/src/templates/rights/rights_edit.html:
```diff
@@ -13,43 +14,56 @@
             {% else %}
-              <p>New {{ new_content_type_created|lower }} content added.</p>
+              <p>
+                {% blocktrans with type=new_content_type_created|lower %}
```
Is `new_content_type_created` already translated? If so, should it be `|lower`, since that may not apply to the language (or in German where I think all nouns have caps)?

Hwesta commented on src/dashboard/src/templates/rights/rights_grants_edit.html:
```diff
@@ -13,37 +14,50 @@
             {% else %}
-              <p>You may now add additional {{ new_content_type_created|lower }} information such as document identifers and notes.</p>
+              <p>
+                {% blocktrans with type=new_content_type_created|lower %}
```
Same comment about `|lower`

sevein replied:
This whole `rights/` component is a pain.

Hwesta commented on src/dashboard/src/templates/transfer/grid.html:
```diff
@@ -99,13 +100,13 @@
       <div class="sip-detail-actions">
-        <a class="btn_show_metadata" href="#" title="Metadata"><span>Micro-Services</span></a>
-        <a class="btn_remove_sip" href="#" title="Remove"><span>Remove</span></a>
+        <a class="btn_show_metadata" href="#" title="{% trans 'Metadatos' %}"><span>{% trans "Metadatos" %}</span></a>
```
This should be left as English.

sevein replied:
Opssss

Hwesta commented on src/dashboard/src/settings/common.py:
```diff
@@ -47,10 +50,31 @@
+    ('fr', _('French')),
+    ('en', _('English')),
+    ('es', _('Spanish')),
+]
```
Is this the final set of languages, or just the ones for testing? If final, why these three, and not others?

sevein replied:
French is a must and I added Spain so I could make my own translations for testing purposes. We can add more later but I wanted to start small to keep it manageable.

Hwesta commented on src/dashboard/src/components/administration/migrations/0007_auto_20161116_1728.py:
```diff
@@ -0,0 +1,19 @@
+    ]
+
+    operations = [
+        migrations.AlterField(
```
This migration should be squashed with the commit that introduced the change before merging.

Hwesta commented on src/MCPClient/lib/clientScripts/emailFailReport.py:
```diff
@@ -29,9 +29,7 @@
-import django
-django.setup()
-# dashboard
+import django; django.setup()
```
Why oneline these? Python generally prefers statements on different lines.

Hwesta commented on src/dashboard/src/main/models.py:
```diff
@@ -403,7 +403,19 @@ class Job(models.Model):
+        (STATUS_EXECUTING_COMMANDS, _('Executing command(s)')),
+        (STATUS_FAILED, _('Failed'))
+    )
+    currentstep = models.IntegerField(db_column='currentStep', choices=STATUS, default=0, blank=False)
```
Don't forget to update `src/dashboard/tests/test_api.py` & related fixtures for these changes.

Hwesta reviewed (commented):
Wow we have a lot of user-facing strings! Code reviewed up to commit "Dashboard: i18njs, import transfer-browser"

sevein committed:
Update gitignore

This commit makes the following changes in .gitignore:

- Add *.mo: gettext binary catalogs
- Add .vscode and .sw[po]: ide-specific entries
- Delete duplicated or unneeded entries (e.g.: mysql_dev.complete)

sevein committed:
Dashboard: add cache config

We're going to need this later in order to cache Django's i18n catalog.

sevein committed:
Dashboard: use static tag in templates

sevein committed:
Dashboard: i18n, enable settings

sevein committed:
Dashboard: i18n, add Transifex config

sevein committed:
Dashboard: i18n, add set_language form

sevein committed:
Dashboard: i18n, annotate tmpls, views, forms...

`templates/tasks.html` has been updated to use `<dl>` and I took the chance to
improve the HTML layout.

sevein committed:
Dashboard: i18n, annotate models

sevein committed:
Dashboard: i18n, import archivematica-fpr-admin w/ pip

sevein committed:
Dashboard: i18njs, setup

Currently not using a cache backend because Archivematica environments don't
include one by default. Note: gearman-job-server is not using memcached, it is
using the builtin in-memory queue instead.

sevein committed:
Dashboard: i18njs, annotate strings with gettext()

TODO: rights_edits.js (big mess!)

sevein committed:
Dashboard: i18njs, update DataTables to v1.10.12

sevein committed:
Dashboard: i18njs, annotate DataTables

Hwesta commented on src/dashboard/src/main/migrations/0030_job_currentstep_choices.py:
```diff
@@ -0,0 +1,41 @@
+
+    # I can't rely on the model registry neither the current model.
+    replacements = {
+        'Awaiting decision': 1,
```
Should this also include unknown: 0?

Hwesta commented on src/dashboard/src/main/migrations/0031_dashboardsetting_scope.py:
```diff
@@ -0,0 +1,238 @@
+
+    # Set new dict and delete previous tuples. Reminder: migrations run inside
+    # a transaction.
+    DashboardSetting.objects.set_dict('upload-qubit_v0.0', new_dict)
```
Will this migration still work if set_dict changes in the future?

sevein replied:
I'm glad you've asked this question because I hadn't thought about it!

I've found two relevant sections in the docs:
- [Historical models](https://docs.djangoproject.com/en/1.8/topics/migrations/#historical-models)
- [Model managers](https://docs.djangoproject.com/en/1.8/topics/migrations/#model-managers)

If I understood properly I can add the attribute `use_in_migrations` (`True`) to `DashboardSettingManager` so the manager becomes part of the migration. However, the manager has to stay around as long as the migration references to the manager. The method `set_dict` should be compatible with all past migrations using it.

More here: https://code.djangoproject.com/ticket/23822.

I could take the two following actions:
- Use `use_in_migrations` (will probably require to regenerate the migration).
- Add a warning to the docs of `set_dict` mention this issue.

Thoughts?

sevein committed:
Dashboard: i18njs, import appraisal-tab

sevein committed:
Dashboard: i18njs, import transfer-browser

sevein committed:
Models: add i18n-choices to Job.currentstep

This commit converts `Job.currentstep` into an `IntegerField` with a set of
`choices` tuples mapped to lazy gettext translations. The database migration
that is included replaces the old values with the corresponding integers. The
affected test fixtures have also been updated.

sevein committed:
DashboardSetting: config dicts for DIP upload settings

This commit introduces a couple of extra methods in the default manager of
DashboardSetting to support table-level sets of simple key: value pairs of
strings. It replaces the models previously available under
`components.administration` used to store DIP upload settings.

For example:

    > DashboardSetting.objects.set_dict('your-setting', {'foo', 'bar'})
    > DashboardSetting.objects.get_dict('your-setting')
    {'foo', 'bar'}

Be aware that both `name` and `value` fields in DashboardSettings are text
fields meaning that all the key: value pairs will be converted into Unicode
strings beforehand.

linkTaskManagerReplacementDicFromChoice has been upgraded to look up these new
configuration dicts after the `execute` column in `StandardTaskConfig`, e.g.
upload-qubit_v0.0, upload-archivesspace_v0.0 or upload-archivistsToolkit_v0.0.

The advatange of this approach is that `StandardTaskConfig` and
`MicroServiceChoiceReplacementDic` become read-only models, facilitating
the upcoming work on workflow data.

Hwesta commented on src/protos/workflow.proto:
```diff
@@ -0,0 +1,107 @@
+            string execute = 1;
+            string arguments = 2;
+            string filterSubdir = 3;
+            string filterFileStart = 4;
```
Looks like this field is actually never used (the value is always null) - may we can just remove it?  Possibly out of scope for this change.

sevein replied:
I think it makes sense to get rid of it. I've done the same with other unused fields IIRC. I will double check before but sure!

Hwesta commented on src/archivematicaCommon/lib/mcpserver_cli.py:
```diff
@@ -0,0 +1,60 @@
+    client = mcpserver.get_client()
+
+    if cmd not in commands:
+        usage(commands)
```
Why not use `argparse` for this sort of checking? It automatically generates usage help, can check that a parameter is in a list, and use subparsers if different commands need different parameters.

sevein replied:
You're right. I will do it. They were conceived for testing purposes but I think they can be handy. If `archivematica-common` becomes its own pypi package, these two could be defined as extra entry points. Eventually I think we should have just one public-facing CLI tool but we're still far from that.

Hwesta commented on src/archivematicaCommon/lib/mcpserver_cli.py:
```diff
@@ -0,0 +1,60 @@
+        except IndexError:
+            sys.exit('Missing parameters (job_uuid, choice_uuid)')
+        approved = client.approve_job(job_uuid, choice_uuid)
+        print("Has the job been approved? {}".format("Yes!" if approved else "No, :("))
```
Clearly this should be `print("Has the job been approved? {}".format("Yes!" if approved else "No, 😞"))` ;)

Hwesta commented on src/MCPServer/lib/gRPCServer.py:
```diff
@@ -0,0 +1,181 @@
+                    break
+            try:
+                match
+            except NameError:
```
I think it's more idiomatic to set `match = None` at the top, and use `if match is None`.

sevein committed:
Dashboard: i18n, makemessages

This commit includes the changes produced by the following commands:

    $ python manage.py makemessages -d django
    $ python manage.py makemessages -d djangojs

Hwesta commented on src/dashboard/src/media/js/jobs.js:
```diff
@@ -433,132 +410,56 @@ var MicroserviceGroupView = Backbone.View.extend({
+          return '/ingest/preview/normalization/' + uuid;
+        // Move to the uploadedDIPs directory
+        } else if (linkId == '2e31580d-1678-474b-83e5-a53d97d150f6' || linkId == 'e3efab02-1860-42dd-a46c-25601251b930') {
+          return '/ingest/preview/dip/' + uuid;
```
Is it possible to use Django's `{% url %}` here?

Hwesta commented on src/archivematicaCommon/lib/workflow.py:
```diff
@@ -0,0 +1,66 @@
+
+        self._init_metadata()
+
+        self.channel = grpc.insecure_channel(self.addr)
```
Are we planning on making this a `secure_channel` later?

sevein replied:
Yes to this but I'm hoping I can simplify as much as possible. If we adopt the sidecar model (i.e. we run something like the linkerd proxy in every host) we could just use the insecure channel to speak to the sidecar and let it take care of the encryption for us. See this article: [transparent TLS with linkerd](https://blog.buoyant.io/2016/03/24/transparent-tls-with-linkerd/). That would make things easier from the perspective of our microservices but also operationally speaking. E.g. we would only have to set up the certificates in the sidecar during deployement (private key, public key, authority cert...), we wouldn't have add extra config parameters to our apps, etc...

Hwesta reviewed (commented):
Looked at 'Models: add i18n-choices to Job.currentstep' to half of 'gRPC: MCPServer and Workflow API'  

A few small comments, but overall I like the direction this is going. Removing the DB as the method of communication and replacing it with something that's actually intended to be spread across different architecture parts makes a lot of sense, and will hopefully make database management easier down the line.

I'm interested in what possibilities other than gRPC exist that might also serve the same purpose. Did you look at any other libraries when considering this?

I'll look at the rest of this PR and artefactual-labs/archivematica-workflow after I get back from vacation.

sevein commented:
> Did you look at any other libraries when considering this?

I briefly looked into other solutions. Apache Thrift is very similar but gRPC is gaining some momentum. The list of open source projects using it is really impressive: etcd, docker, mesos, kubernetes... gRPC is interesting in that it was designed to support multiple data formats, protobuf is the default.

Hwesta commented on src/archivematicaCommon/lib/workflow.py:
```diff
@@ -0,0 +1,66 @@
+class ServerUnavailableError(Error):
+    def __init__(self, code):
+        self.code = code
+        self.message = 'server unavailable, code ' + self.code
```
Can't concatenate a str & int - `self.message = 'server unavailable, code %s' % (self.code,)` will work.  Should probably do the same precaution above.
